test/run_pass/test_stream.js

```diff
@@ -18,60 +18,152 @@ var Readable = require('stream').Readable;
   e += 'e';
 });
 
 
-readable.pause();
-readable.push('abcde');
-readable.push('12345');
+readable1.pause();
+readable1.push('abcde');
+readable1.push('12345');
 assert.equal(d, '');
 assert.equal(e, '');
 
-readable.resume();
+readable1.resume();
 assert.equal(d, 'abcde12345');
 assert.equal(e, '');
 
-readable.push('a');
-readable.push('1');
-readable.push('b');
-readable.push('2');
+readable1.push('a');
+readable1.push('1');
+readable1.push('b');
+readable1.push('2');
 assert.equal(d, 'abcde12345a1b2');
 assert.equal(e, '');
 
-readable.pause();
+assert.equal(readable1.isPaused(), false);
+readable1.pause();
 assert.equal(d, 'abcde12345a1b2');
 assert.equal(e, '');
+assert.equal(readable1.isPaused(), true);
 
-readable.push('c');
-readable.push('3');
-readable.push('d');
-readable.push('4');
+// Pause the readable again. This should do nothing.
+readable1.pause();
+assert.equal(readable1.isPaused(), true);
+
+readable1.push('c');
+readable1.push('3');
+readable1.push('d');
+readable1.push('4');
 assert.equal(d, 'abcde12345a1b2');
 assert.equal(e, '');
 
-readable.resume();
+readable1.resume();
 assert.equal(d, 'abcde12345a1b2c3d4');
 assert.equal(e, '');
 
-readable.push(null);
+readable1.push(null);
 assert.equal(d, 'abcde12345a1b2c3d4');
 assert.equal(e, 'e');
 
-readable.push('push after eof');
+readable1.push('push after eof');
 assert.equal(d, 'abcde12345a1b2c3d4');
 assert.equal(e, 'e.');
+
+
+// Create a readable stream without the new keyword.
+var readable2 = Readable({encoding: 'utf8'});
+
+// Push an invalid chunk into it.
+assert.throws(function() {
+  readable2.push(undefined);
+}, TypeError);
+
+assert.throws(function() {
+  readable2.push(5001);
+}, TypeError);
+
+assert.throws(function() {
+  readable2.push(5001.5);
+}, TypeError);
+
+assert.throws(function() {
+  readable2.push([]);
+}, TypeError);
+
+assert.throws(function() {
+  readable2.push([5001, 'string']);
+}, TypeError);
+
+assert.throws(function() {
+  readable2.push({});
+}, TypeError);
+
+assert.throws(function() {
+  readable2.push({obj: 'string', second: 5001});
+}, TypeError);
+
+// Read with irregular parameters from an empty stream.
+assert.equal(readable2.read(-2), null);
+assert.equal(readable2.read(0), null);
+
+readable2.push('qwerty');
+assert.equal(readable2.read(6), 'qwerty');
+
+// Try to read less data than we have.
+readable2.push('new-data');
+assert.throws(function() {
+  readable2.read(1);
+}, Error);
+
+var readable3 = new Readable();
+var readable3End = false;
+var readable3Data = '';
+
+readable3.on('data', function(data) {
+  readable3Data += data.toString();
+});
+
+readable3.on('end', function() {
+  readable3End = true;
+});
+
+var str = 'test';
+readable3.push(str);
+readable3.pause();
+
+// Push something to the stream while paused.
+readable3.push(str);
+readable3.push(null);
+
+// Resume the stream after 1s,
+// the pushed str and null shuld be processed after this.
+setTimeout(function() {
+  readable3.resume();
+}, 1000);
```

Comment:
I'm not sure that `pause` and `resume` work well with this test. How about the following?

```js
var paused = false;
var str = 'test';

readable3.on('data', function(data) {
  assert.equal(paused, true);
  assert.equal(data, str);
});

readable3.pause();
readable3.push(str);
readable3.push(null);

setTimeout(function() {
  paused = true;
  readable3.resume();
}, 1000);
```